criterion-measurement/src/Criterion/Measurement.hs

```diff
@@ -390,6 +404,31 @@ secs k
                | t >= 1e1  = printf "%.2f %s" t u
                | otherwise = printf "%.3f %s" t u
 
+#if defined(__GHCJS__)
+-- atm, cabal haddock --ghcjs fails :(
+-- So, not trying to make haddocks for GHCJS variants.
+initializeTime :: IO ()
+initializeTime = return ()
+
+-- | In browser, we don't know cycles. Always returns @0@.
+getCycles :: IO Word64
+getCycles = return 0
+
+-- | In browser we don't know about a CPU. Always returns @0@.
+getCPUTime :: IO Double
+getCPUTime = return 0
+
+-- Performance API is in all major browsers: https://developer.mozilla.org/en-US/docs/Web/API/Performance
+-- and in NodeJS since 8.5.0: https://nodejs.org/api/perf_hooks.html#perf_hooks_performance_timing_api
+--
+-- Examples of GHCJS FFI: https://github.com/ghcjs/ghcjs/wiki/A-few-examples-of-Foreign-Function-Interface
+--
+-- performance.now() returns *milliseconds*. So we divide them to get seconds.
+foreign import javascript unsafe
+  "if (typeof window === 'object' && typeof window.performance === 'object') { $r = window.performance.now() / 1000; } else if (typeof require === 'function') { $r = require('perf_hooks').performance.now() / 1000; } else { $r = 0.0; }"
```

Comment:
This is... a lot of JavaScript to put on a single line. Is there any way to factor this out somehow?

Reply:
In separate file? I don't know how, maybe. If we had multiline string literals, it would be ok; it's not that long after all.
Comment above hints what happens there.